From: Tim Chen <tim.c.chen@linux.intel.com>
Date: Mon, 18 Dec 2017 10:37:31 -0800
Subject: x86/microcode: Recheck IBRS and IBPB feature on microcode reload
Patch-mainline: submitted on 2018/1/9
References: bnc#1068032 CVE-2017-5715

On new microcode write, check whether IBRS and IBPB features
are present by rescanning scattered CPU features.

Signed-off-by: Tim Chen <tim.c.chen@linux.intel.com>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/kernel/cpu/microcode/core.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kernel/cpu/microcode/core.c b/arch/x86/kernel/cpu/microcode/core.c
index c4fa4a85d4cb..6048612364e3 100644
--- a/arch/x86/kernel/cpu/microcode/core.c
+++ b/arch/x86/kernel/cpu/microcode/core.c
@@ -444,6 +444,11 @@ static ssize_t microcode_write(struct file *file, const char __user *buf,
 	if (ret > 0)
 		perf_check_microcode();
 
+	/* check spec_ctrl capabilities */
+	mutex_lock(&spec_ctrl_mutex);
+	init_scattered_cpuid_features(&boot_cpu_data);
+	mutex_unlock(&spec_ctrl_mutex);
+
 	mutex_unlock(&microcode_mutex);
 	put_online_cpus();
 
-- 
2.15.1

